staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -504,11 +555,19 @@ func (r *crdHandler) updateCustomResourceDefinition(oldObj, newObj interface{})
 
 	if oldCRD.UID != newCRD.UID {
 		r.removeStorage_locked(oldCRD.UID)
+		if utilfeature.DefaultFeatureGate.Enabled(features.StorageVersionAPI) &&
+			utilfeature.DefaultFeatureGate.Enabled(features.APIServerIdentity) {
+			r.storageVersionManager.TeardownFor(oldCRD.UID)
+		}
 	}
 
 	storageMap := r.customStorage.Load().(crdStorageMap)
 	oldInfo, found := storageMap[newCRD.UID]
 	if !found {
+		if utilfeature.DefaultFeatureGate.Enabled(features.StorageVersionAPI) &&
+			utilfeature.DefaultFeatureGate.Enabled(features.APIServerIdentity) {
+			r.storageVersionManager.UpdateStorageVersionFor(newCRD, nil, nil, nil)
```

Comment:
maybe I missed it, do we check anywhere that the update is necessary? Would feel better to have only very specific cases where we queue an update.

Reply:
Line 515 below is new, right? That would cause the skip because the code above only fires when `!found` is true and this is never the case on normal updates? Maybe this is worth a oneliner comment here.

Reply:
this is case 2), and line 515 is case 4) in https://github.com/kubernetes/kubernetes/pull/96403#discussion_r558146492. This happens when a CRD gets updated before any CR request hits the server. 

re-organized the code and added comments: we skip storage version update if an old-and-exact-same storage exists, in which case we know we have done the same storage version update already; otherwise we do a storage version update with the latest info in the watch event.

Reply:
:+1: